servicelayerfunctionalities.cc

```diff
@@ -0,0 +1,328 @@
+  {
+    std::lock_guard<std::mutex> lock(mymutex_);
+    auto fromget = clientKey.get("chirps_");
+    if (fromget.size() == 0) {
+      nextChirpID = std::to_string(0);
+    } else {
+      int currchirpid = std::stoi(fromget[0]);
+      currchirpid++;
+      nextChirpID = std::to_string(currchirpid);
+    }
+    clientKey.put("chirps_", nextChirpID);
+  }
+  //Get User
+  chirp::User user;
+  std::vector <std::string> fromget = clientKey.get(request->username());
+  if (fromget.size() == 0) {
+    return Status::CANCELLED;
+  }
+  std::string getValue = fromget[0];
+  user.ParseFromString(getValue);
+  std::string userKey;
+  
+  //Add new Chirp to chirp::User
+  std::string value; 
+  {
+    chirp::Chirp* message = user.add_chirps();
+    message->set_username(request->username());
+    message->set_text(request->text());
+    message->set_id(nextChirpID);  
+    chirp::Timestamp* timestamp = message->mutable_timestamp();
+    timestamp->set_seconds(static_cast<int64_t>(seconds));
+    timestamp->set_useconds(microseconds_since_epoch);
+    message->set_parent_id(request->parent_id());
+    message->SerializeToString(&value);
+    
+    setChirpReply(message, response);
+  }
+
+  user.SerializeToString(&userKey);
+  clientKey.put(request->username(),userKey); //Add Chirp to chirp::User
+  clientKey.put("chirp"+nextChirpID,value);//Add chirp::Chirp to database
+  
+  //If its a reply to an existing chirp
+  if (request->parent_id() != "") { 
+    std::string value2; 
+    {
+      std::vector <std::string> fromget = clientKey.get("reply"+request->parent_id());
+
+      chirp::ChirpReplies replies;
+      
+      //If chirp id is not found, use the 'replies' as a new ChirpReplies and set parent_id()
+      if (fromget.size() == 0) { 
+        replies.set_parent_id(request->parent_id());
+      } else { //If found, parse from existing ChirpReplies
+        replies.ParseFromString(fromget[0]);
+      }
+      chirp::Chirp* reply = replies.add_chirps();
+      reply->set_username(request->username());
+      reply->set_text(request->text());
+      reply->set_id(nextChirpID);
+      reply->set_parent_id(request->parent_id());
+      chirp::Timestamp* timestamp = reply->mutable_timestamp();
+      timestamp->set_seconds(static_cast<int64_t>(seconds));
+      timestamp->set_useconds(microseconds_since_epoch);
+      replies.SerializeToString(&value2);
+    }
+    clientKey.put("reply"+request->parent_id(),value2); //Add reply<ID> to backend
+  }
+  
+  return Status::OK;
+}
+
+Status ServerForCommandLineClient::read(ServerContext* context, const ReadRequest* request, ReadReply* response) {
+  /*
+    Tested: Able to get all chirps and it's replies and send it back via ReadReply
+  */
+  ClientForKeyValueStore clientKey(grpc::CreateChannel("localhost:50000", grpc::InsecureChannelCredentials()));
+
+  std::vector <std::string> fromget = clientKey.get("chirp"+request->chirp_id());
+  std::vector <chirp::Chirp> chirpToSend; //For sorting by timestamp
```

Comment:
Variable names should be lowercase w/ underscores as appropriate

Reply:
done